js/src/DappRequests/store.js

```diff
@@ -41,131 +38,137 @@ export default class Store {
   }
 
   @computed get hasRequests () {
-    return this.requests.length !== 0;
+    return Object.keys(this.requests).length !== 0;
   }
 
-  @computed get squashedRequests () {
-    const duplicates = {};
+  @computed get groupedRequests () {
+    // Group by appId on top level, and by methodGroup on 2nd level
+    return Object.values(this.requests).reduce((accumulator, request) => {
+      const { data: { method, params, token } } = request;
+      const appId = this.tokens[token];
 
-    return this.requests.filter(({ request: { data: { method, token } } }) => {
-      const section = this.getFilteredSectionName(method);
-      const id = `${token}:${section}`;
+      accumulator[appId] = accumulator[appId] || {};
 
-      if (!duplicates[id]) {
-        duplicates[id] = true;
-        return true;
-      }
+      const methodGroup = methodGroupFromMethod[method || params[0]]; // Get the methodGroup the current request belongs to
 
-      return false;
-    });
+      accumulator[appId][methodGroup] = accumulator[appId][methodGroup] || [];
+      accumulator[appId][methodGroup].push(request);
+      return accumulator;
+    }, {});
   }
 
-  @action createToken = (appId) => {
+  @action createToken = appId => {
     const token = sha3(`${appId}:${Date.now()}`);
 
     this.tokens = Object.assign({}, this.tokens, {
       [token]: appId
     });
 
     return token;
-  }
-
-  @action removeRequest = (_queueId) => {
-    this.requests = this.requests.filter(({ queueId }) => queueId !== _queueId);
-    delete this.sources[_queueId];
-  }
-
-  @action queueRequest = (request) => {
-    const { data, origin, source } = request;
-    const appId = this.tokens[data.token];
-
-    let queueId = ++nextQueueId;
+  };
 
-    this.sources[queueId] = source;
-    this.requests = this.requests.concat([{
-      appId,
-      queueId,
-      request: {
-        data,
-        origin
-      }
-    }]);
-  }
+  @action queueRequest = request => {
+    const { data: { method, token } } = request;
+    const appId = this.tokens[token];
 
-  @action addTokenPermission = (method, token) => {
-    const id = `${method}:${this.tokens[token]}`;
-
-    this.permissions = Object.assign({}, this.permissions, {
-      [id]: true
-    });
-    this.savePermissions();
-  }
+    // Add a new request in this.requests
+    this.requests = {
```

Comment:
Why copy `this.requests` over to a new object?

Reply:
Oh it's because of `mobx`, right? We need to assign a new object to refresh the store? Makes sense.

Reply:
Exactly. Since we're dynamically adding new properties to the observable object, I think that's the best way to update the store